File: ios/SCGPushSDK/SCGPushSDK/SCGPush.m
//
//  SCGPush.m
//  SCGPushSDK
//
//  Created by Angel Terziev on 3/13/17.
//  Copyright © 2017 Syniverse. All rights reserved.
//

#import "SCGPush.h"
#import "HttpRedirectDecisionMaker.h"

#import "SCGPushCoreDataManager.h"

#import <MobileCoreServices/UTCoreTypes.h>

static SCGPush *_sharedInstance = nil;

@interface SCGPush()
@property (nonatomic, strong) HttpRedirectDecisionMaker* redirectDecisionMaker;
@property (nonatomic, strong) SCGPushCoreDataManager* coreDataManager;

@property (nonatomic, copy, nonnull) NSString* accessTokenInternal;
@property (nonatomic, copy, nonnull) NSString* callbackURIInternal;
@property (nonatomic, copy, nonnull) NSString* appIDInternal;

@end

@implementation SCGPush

+ (instancetype) sharedInstanceWithDelegate: (id<SCGPushDelegate>) delegate
{
    // 2
    static dispatch_once_t oncePredicate;
    
    // 3
    dispatch_once(&oncePredicate, ^{
        _sharedInstance = [[SCGPush alloc] initWithDelegate: delegate];
    });
    
    return _sharedInstance;
}

+ (instancetype) sharedInstance
{
    return [[self class] sharedInstanceWithDelegate: nil];
}

- (instancetype) initWithDelegate: (id<SCGPushDelegate>) delegate
{
    if (nil != (self = [super init])) {
        NSLog(@"Debug: [SCGPush] Initialzing  <%p>", self);
        _delegate = delegate;
        
        self.accessTokenInternal = @"";
        self.appIDInternal = @"";
        self.callbackURIInternal = @"";
    }
    
    return self;
}

+ (instancetype _Nonnull) startWithAccessToken: (NSString* _Nonnull) accessToken
                                         appId: (NSString* _Nonnull) appId
                                   callbackUri: (NSString* _Nonnull) callbackUri
                                      delegate: (id<SCGPushDelegate> _Nullable) delegate
{
    SCGPush* scgPush = [[self class] sharedInstanceWithDelegate: delegate];
    
    scgPush.accessToken = accessToken;
    scgPush.appID = appId;
    scgPush.callbackURI = callbackUri;
    
    [scgPush initializeInbox];
    
    return scgPush;
}


//MARK: - Push Token

- (void) registerPushToken:( NSString* _Nonnull) pushToken
     withCompletionHandler:( void (^)(NSString * _Nullable token)) completionBlock
              failureBlock:( void (^)(NSError * _Nullable error)) failureBlock
{
    if (nil == pushToken) {
        if (failureBlock) {
            NSError* error = [NSError errorWithDomain: @"SCGPush" code: 1 userInfo: nil];
            failureBlock(error);
        }
        
        return;
    }
    
    NSURLSessionConfiguration* configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
    NSURLSession* session = [NSURLSession sessionWithConfiguration: configuration];
    NSString* urlString = [NSString stringWithFormat: @"%@/push_tokens/register",
                           self.callbackURI];
    NSURL* url = [NSURL URLWithString: urlString];
    if (nil == url) {
        if (failureBlock) {
            NSError* error = [NSError errorWithDomain: @"SCGPush" code: 2 userInfo: nil];
            failureBlock(error);
        }
        return;
    }
    
    NSLog(@"Debug: [SCGPush] URL: %@", url.absoluteString);
    
    NSMutableURLRequest* request = [NSMutableURLRequest requestWithURL: url];
    request.HTTPMethod = @"POST";
    request.timeoutInterval = 30;

    [request addValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    NSString* bearer = [NSString stringWithFormat:@"Bearer %@", self.accessToken];
    [request addValue: bearer forHTTPHeaderField:@"Authorization"];
    NSError* jsonError = nil;
    NSDictionary* params = @{@"app_id": self.appID,
                             @"type": @"APN",
                             @"token": pushToken};
    NSData* jsonData = [NSJSONSerialization dataWithJSONObject:params options:0 error: &jsonError];
    if (nil == jsonData) {
        if (failureBlock) {
            failureBlock(jsonError);
        }
        return;
    }
    
    request.HTTPBody = jsonData;

    NSURLSessionDataTask* dataTask =
    [session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        NSHTTPURLResponse* httpResponse = (NSHTTPURLResponse*) response;
        if (error != nil) {
            if (failureBlock) {
                failureBlock(error);
            }
            return;
        }

        switch (httpResponse.statusCode) {
            case 200:
            case 204:
            {
                if (completionBlock != nil) {
                    completionBlock(pushToken);
                }
            }
                break;
            default:
            {
                if (failureBlock) {
                    NSError* error = [NSError errorWithDomain: @"SCGPush" code: httpResponse.statusCode userInfo: nil];
                    failureBlock(error);
                }
            }
                break;
                
        }
    }];

    [dataTask resume];
}

// MARK: - Report Status
- (NSString*) messageStateToString: (MessageState) state {

    NSString* stateAsString = nil;
    switch (state) {
        case MessageStateDelivered:
            stateAsString = @"DELIVERED";
            break;
        case MessageStateRequested:
            stateAsString = @"MEDIA_REQUESTED";
            break;
        case MessageStateRead:
            stateAsString = @"READ";
            break;
        case MessageStateClicked:
            stateAsString = @"CLICKTHRU";
            break;
        case MessageStateConverted:
            stateAsString = @"CONVERTED";
            break;
        default:
            break;
    }
    
    return stateAsString;
}

- (void) reportStatusWithMessageId: ( NSString* _Nonnull) messageId
                   andMessageState: ( MessageState ) state
                   completionBlock: ( void(^ _Nullable)()    ) completionBlock
                     failureBlock : ( void(^ _Nullable) (NSError* _Nullable error)) failureBlock
{

    NSString* stringifiedMessageState = [self messageStateToString: state];
    if (nil == stringifiedMessageState) {
        if (failureBlock) {
            NSError* error = [NSError errorWithDomain: @"SCGPush" code: 1 userInfo: nil];
            failureBlock(error);
        }
        return;
    }
    
    NSURLSessionConfiguration* configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
    NSURLSession* session = [NSURLSession sessionWithConfiguration: configuration];
    NSString* urlString = [NSString stringWithFormat: @"%@/messages/%@/confirm/%@",
                           self.callbackURI,
                           messageId,
                           stringifiedMessageState];

    NSURL* url = [NSURL URLWithString: urlString];
    if (nil == url) {
        if (failureBlock) {
            NSError* error = [NSError errorWithDomain: @"SCGPush" code: 2 userInfo: nil];
            failureBlock(error);
        }
        return;
    }
    
    NSLog(@"Debug: [SCGPush] URL: %@", url.absoluteString);
    
    NSMutableURLRequest* request = [NSMutableURLRequest requestWithURL: url];
    request.HTTPMethod = @"POST";
    request.timeoutInterval = 30;
    
    [request addValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    NSString* bearer = [NSString stringWithFormat:@"Bearer %@", self.accessToken];
    [request addValue: bearer forHTTPHeaderField:@"Authorization"];

    NSURLSessionDataTask* dataTask =
    [session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        NSHTTPURLResponse* httpResponse = (NSHTTPURLResponse*) response;
        if (error != nil) {
            if (failureBlock) {
                failureBlock(error);
            }
            return;
        }
        
        switch (httpResponse.statusCode) {
            case 200:
            case 204:
            {
                if (completionBlock != nil) {
                    completionBlock();
                }
            }
                break;
            default:
            {
                if (failureBlock) {
                    NSError* error = [NSError errorWithDomain: @"SCGPush" code: httpResponse.statusCode userInfo: nil];
                    failureBlock(error);
                }
            }
                break;
        }
    }];
    
    [dataTask resume];
}

// MARK: Resolve Tracked Link
- (void) resolveTrackedLink:(NSString* _Nonnull) urlString
{
    NSURLSessionConfiguration* configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
    NSURLSession* session = [NSURLSession sessionWithConfiguration:configuration delegate:self.redirectDecisionMaker delegateQueue:nil];
    NSURL* url = [NSURL URLWithString: urlString];

    NSLog(@"Debug: [SCGPush] URL: %@", url.absoluteString);
    
    NSMutableURLRequest* request = [NSMutableURLRequest requestWithURL: url];
    request.HTTPMethod = @"HEAD";
    request.timeoutInterval = 30;
    
    NSURLSessionDataTask* dataTask =
    [session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        NSHTTPURLResponse* httpResponse = (NSHTTPURLResponse*) response;
        if (error != nil) {
            return;
        }
        
        switch (httpResponse.statusCode) {
            case 300:
            case 301:
            case 307:
            {
                if (self.delegate != nil) {
                    
                    NSString* redirectLocation = httpResponse.allHeaderFields[@"Location"];
                    if (nil != redirectLocation) {
                        [self.delegate resolveTrackedLinkDidSuccess:redirectLocation withrequest:request];
                    } else {
                        [self.delegate resolveTrackedLinkHasNotRedirect:request];
                    }
                }
            }
                break;
            default:
            {
                if (self.delegate != nil) {
                    [self.delegate resolveTrackedLinkHasNotRedirect:request];
                }
            }
                break;
        }
    }];
    
    [dataTask resume];

}

// MARK: - Load Attachment

- (NSString*) translateContentTypeHeader: (NSHTTPURLResponse* _Nonnull) httpResponse
{
    NSString* translatedContentType = @"";
    NSString* contentType = httpResponse.allHeaderFields[@"Content-Type"];
    
    if (contentType != nil) {
        if (NSOrderedSame == [contentType caseInsensitiveCompare: @"video/mpeg"]) {
            translatedContentType = (NSString*) kUTTypeMPEG4;
        } else if (NSOrderedSame == [contentType caseInsensitiveCompare: @"video/mp4"]) {
            translatedContentType = (NSString*) kUTTypeMPEG4;
        } else if (NSOrderedSame == [contentType caseInsensitiveCompare: @"video/webm"]) {
            translatedContentType = (NSString*) kUTTypeVideo;
        } else if (NSOrderedSame == [contentType caseInsensitiveCompare: @"video/ogg"]) {
            translatedContentType = (NSString*) kUTTypeVideo;
        } else if (NSOrderedSame == [contentType caseInsensitiveCompare: @"audio/ogg"]) {
            translatedContentType = (NSString*) kUTTypeAudio;
        } else if (NSOrderedSame == [contentType caseInsensitiveCompare: @"audio/webm"]) {
            translatedContentType = (NSString*) kUTTypeAudio;
        } else if (NSOrderedSame == [contentType caseInsensitiveCompare: @"audio/mpeg"]) {
            translatedContentType = (NSString*) kUTTypeAudio;
        } else if (NSOrderedSame == [contentType caseInsensitiveCompare: @"audio/mp3"]) {
            translatedContentType = (NSString*) kUTTypeMP3;
        } else if (NSOrderedSame == [contentType caseInsensitiveCompare: @"image/gif"]) {
            translatedContentType = (NSString*) kUTTypeGIF;
        } else if (NSOrderedSame == [contentType caseInsensitiveCompare: @"image/png"]) {
            translatedContentType = (NSString*) kUTTypePNG;
        } else if (NSOrderedSame == [contentType caseInsensitiveCompare: @"image/jpeg"]) {
            translatedContentType = (NSString*) kUTTypeJPEG;
        }
    }
    
    return translatedContentType;
    
}

- (void) loadAttachmentWithMessageId:(NSString* _Nonnull) messageId
                     andAttachmentId:(NSString* _Nonnull) attachmentId
                     completionBlock:(void(^_Nullable)(NSURL* _Nonnull contentUrl, NSString* _Nonnull contentType))completionBlock
                        failureBlock:(void(^_Nullable)(NSError* _Nullable error))failureBlock
{
    NSLog(@"Debug: [SCGPush] '<%p>', will load attachment '%@' for message '%@'",
          self,
          attachmentId,
          messageId);
    NSURLSessionConfiguration* configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
    NSURLSession* session = [NSURLSession sessionWithConfiguration:configuration];
    NSString* urlString = [NSString stringWithFormat: @"%@/attachment/%@/%@",
                           self.callbackURI,
                           messageId,
                           attachmentId];
    NSURL* url = [NSURL URLWithString: urlString];
    
    NSLog(@"Debug: [SCGPush] URL: %@", url.absoluteString);
    NSMutableURLRequest* request = [NSMutableURLRequest requestWithURL: url];
    request.HTTPMethod = @"GET";
    request.timeoutInterval = 30;

    NSString* bearer = [NSString stringWithFormat:@"Bearer %@", self.accessToken];
    [request addValue: bearer forHTTPHeaderField:@"Authorization"];

    NSURLSessionDownloadTask* downloadTask =
    [session downloadTaskWithRequest:request completionHandler:^(NSURL * _Nullable location,
                                                                 NSURLResponse * _Nullable response,
                                                                 NSError * _Nullable error) {
        NSHTTPURLResponse* httpResponse = (NSHTTPURLResponse*) response;
        if (error != nil) {
            if (failureBlock) {
                failureBlock(error);
            }
            return;
        }
        
        NSURL* contentUrl;
        NSString* contentType = [self translateContentTypeHeader: httpResponse];
        
        if (nil != location ) {
            // Move temporary file to remove .tmp extension
            NSString* tmpDirectory = NSTemporaryDirectory();
            NSString*  tmpFile = [[@"file://" stringByAppendingString:tmpDirectory] stringByAppendingString:url.lastPathComponent];
            NSURL* tmpUrl = [NSURL URLWithString:tmpFile];
            if ([[NSFileManager defaultManager] fileExistsAtPath:tmpUrl.path isDirectory:nil])
                [[NSFileManager defaultManager] removeItemAtURL:tmpUrl error:nil];
            
            [[NSFileManager defaultManager] moveItemAtURL:location toURL:tmpUrl error:nil];
            contentUrl = tmpUrl;
        }

        switch (httpResponse.statusCode) {
            case 200:
            case 204:
            {
                if (completionBlock != nil) {
                    completionBlock(contentUrl, contentType);
                }
            }
                break;
            default:
            {
                if (failureBlock) {
                    NSError* error = [NSError errorWithDomain: @"SCGPush" code: httpResponse.statusCode userInfo: nil];
                    failureBlock(error);
                }
            }
                break;
        }
    }];
    
    [downloadTask resume];
}

// MARK: - Save Device Token
- (void) saveDeviceToken: (NSString* _Nonnull) token
{
}

- (void) saveDeviceTokenData: (NSData* _Nonnull) tokenData
{
}

//MARK: - PushInbox
- (void) initializeInbox
{
    self.coreDataManager = [SCGPushCoreDataManager sharedInstance];
}

- (BOOL) pushToInbox: (NSDictionary* _Nonnull) payload
{
    BOOL fOk = NO;
    id identifier = payload[@"scg-message-id"];
    id body = payload[@"body"];
    id deepLink = payload[@"deep_link"];
    id attachmentId = payload[@"scg-attachment-id"];
    id showNotification = payload[@"show-notification"];
    
    if (nil == identifier || ![identifier isKindOfClass: [NSString class]])
        return fOk;
    
    NSString* messageId = (NSString*) identifier;
    NSString* messageBody = @"";
    
    if (nil != body && [body isKindOfClass:[NSString class]]) {
        messageBody = (NSString*) body;
    }
    
    SCGPushMessage* message =
    [[SCGPushMessage alloc] initWithId: messageId
                           dateCreated: [NSDate date]
                               andBody: messageBody];
    
    if (!message)
        return fOk;
    
    if (deepLink && [deepLink isKindOfClass:[NSString class]]) {
        message.deepLink = (NSString*) deepLink;
    }
    if (attachmentId && [attachmentId isKindOfClass:[NSString class]]) {
        message.attachmentId = (NSString*) attachmentId;
    }
    
    if (showNotification && [showNotification isKindOfClass:[NSNumber class]]) {
        message.showNotification = [showNotification boolValue];
    }
    
    fOk = [self.coreDataManager addNewMessage: message];
    
    return fOk;
}

//MARK: - accessors
- (HttpRedirectDecisionMaker *)redirectDecisionMaker {

    if (nil == _redirectDecisionMaker) {
        _redirectDecisionMaker = [[HttpRedirectDecisionMaker alloc] initWithPreventRedirect: YES];
    }
    
    return _redirectDecisionMaker;
}

-(void)setAccessToken:(NSString *)newValue {
    @synchronized (self) {
        if ([newValue length]) {
            NSLog(@"Debug: [SCGPush] set accessToken %@", newValue);
            self.accessTokenInternal = [newValue copy];
        } else{
            NSLog(@"Debug: [SCGPush] clear accessToken");
            self.accessTokenInternal = [@"" copy];
        }
    }
}

-(NSString *)accessToken {
    @synchronized (self) {
        return self.accessTokenInternal;
    }
}

-(void)setAppID:(NSString *)newValue {
    @synchronized (self) {
        if ([newValue length]) {
            NSLog(@"Debug: [SCGPush] set appID %@", newValue);
            self.appIDInternal = [newValue copy];
        } else{
            NSLog(@"Debug: [SCGPush] clear appID");
            self.appIDInternal = [@"" copy];
        }
    }
}

-(NSString *)appID {
    @synchronized (self) {
        return self.appIDInternal;
    }
}

-(void)setCallbackURI:(NSString *)newValue{
    @synchronized (self) {
        if ([newValue length]) {
            NSLog(@"Debug: [SCGPush] set callbackURI %@", newValue);
            self.callbackURIInternal = [newValue copy];
        } else{
            NSLog(@"Debug: [SCGPush] clear callbackURI");
            self.callbackURIInternal = [@"" copy];
        }
    }
    
}

-(NSString *)callbackURI {
    @synchronized (self) {
        return self.callbackURIInternal;
    }
}
@end

File: ios/SCGPush/SCGPush/SCGPush.swift
//
//  SCGPush.swift
//  SCGPush
//
//  Created by Slav Sarafski on 6/12/16.
//  Copyright © 2016 Spirit Invoker. All rights reserved.
//

import UIKit
import MobileCoreServices

open class SCGPush: NSObject {
    
    // PRIVATE VARIABLES
    fileprivate var _accessToken:String = ""
    
    // PUBLIC VARIABLES
    open var accessToken:String
    {
        set {
            let defaults = UserDefaults.standard
            defaults.set(newValue, forKey: ("scg-access-token-dont-replace-this-default"))
            if groupBundle != "" {
                if let groupDefault = UserDefaults(suiteName: groupBundle) {
                    groupDefault.set(newValue, forKey: ("scg-access-token-dont-replace-this-default"))
                }
            }
            _accessToken = newValue
        }
        get {
            _accessToken = ""
            let defaults = UserDefaults.standard
            if (defaults.string(forKey: "scg-access-token-dont-replace-this-default") != nil) {
                _accessToken = defaults.string(forKey: "scg-access-token-dont-replace-this-default")!
            } else {
                if let groupDefault = UserDefaults(suiteName: groupBundle) {
                    if let token = groupDefault.string(forKey: "scg-access-token-dont-replace-this-default") {
                        _accessToken = token
                    }
                }
            }
            return _accessToken
        }
    }
    
    open var callbackURI:String
    {
        set {
            let defaults = UserDefaults.standard
            defaults.set(newValue, forKey: ("scg-callback-uri-dont-replace-this-default"))
            if groupBundle != "" {
                if let groupDefault = UserDefaults(suiteName: groupBundle) {
                    groupDefault.set(newValue, forKey: ("scg-callback-uri-dont-replace-this-default"))
                }
            }
        }
        get {
            let defaults = UserDefaults.standard
            if (defaults.string(forKey: "scg-callback-uri-dont-replace-this-default") != nil) {
                return defaults.string(forKey: "scg-callback-uri-dont-replace-this-default")!
            }
            else {
                if let groupDefault = UserDefaults(suiteName: groupBundle) {
                    if let uri = groupDefault.string(forKey: "scg-callback-uri-dont-replace-this-default") {
                        return uri
                    }
                }
            }
            return ""
        }
    }
    
    open var appID:String = ""
    
    open var groupBundle:String = ""
    
    // PRIVATE VARIABLES
    fileprivate let tokenType = "APN"
    
    open static let instance = SCGPush()
    
    public override init (){
        
    }
    
    open func registerPushToken(deviceTokeData data:Data, completionBlock: (() -> Void)? = nil, failureBlock : ((Error?) -> ())? = nil) {
        
        let tokenChars = (data as NSData).bytes.bindMemory(to: CChar.self, capacity: data.count)
        var pushToken = ""
        for i in 0..<data.count {
            pushToken += String(format: "%02.2hhx", arguments: [tokenChars[i]])
            
        }
        
        registerPushToken(pushToken, completionBlock: completionBlock, failureBlock: failureBlock)
    }
    
    open func registerPushToken(_ deviceToken:String, completionBlock: (() -> Void)? = nil, failureBlock : ((Error?) -> ())? = nil)
    {
        saveDeviceToken(deviceToken: deviceToken as NSString)
        
        registerPushToken(completionBlock, failureBlock: failureBlock)
    }
    
    open func registerPushToken(_ completionBlock: (() -> Void)? = nil, failureBlock : ((Error?) -> ())? = nil)
    {
        let defaults = UserDefaults.standard
        if (defaults.string(forKey: "scg-push-token-dont-replace-this-default") == nil) {
            return
        }
        
        let deviceToken:String = defaults.string(forKey: "scg-push-token-dont-replace-this-default")!
        
        
        let params = ["app_id":appID as AnyObject,
                      "type":tokenType as AnyObject,
                      "token":deviceToken as AnyObject] as Dictionary<String, AnyObject>
        
        let configuration = URLSessionConfiguration.default
        let session = URLSession(configuration: configuration)
        
        let urlString = "\(callbackURI)/push_tokens/register"
        let url = URL(string: urlString)
        
        var request = URLRequest(url: url!)
        request.httpMethod = "POST"
        request.timeoutInterval = 30
        
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.addValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        request.httpBody  = try! JSONSerialization.data(withJSONObject: params, options: [])
        
        let dataTask = session.dataTask(with: request, completionHandler: {
            (data: Data?, response: URLResponse?, error: Error?) -> Void in
            
            guard let httpResponse = response as? HTTPURLResponse, let _ = data
                else {
                    //print("error: not a valid http response")
                    if (failureBlock != nil) {
                        failureBlock! (error)
                    }
                    return
            }
            
            switch (httpResponse.statusCode)
            {
            case 200:
                if (completionBlock != nil) {
                    completionBlock! ()
                }
            case 204:
                if (completionBlock != nil) {
                    completionBlock! ()
                }
                
            default:
                if (failureBlock != nil) {
                    let errorSend = NSError(domain: (httpResponse.url?.absoluteString)!, code: httpResponse.statusCode, userInfo: nil)
                    failureBlock! (errorSend)
                }
            }
        })        

        dataTask.resume()
    }

    open func unregisterPushToken(_ completionBlock: (() -> Void)? = nil, failureBlock : ((Error?) -> ())? = nil)
    {
        let defaults = UserDefaults.standard
        let pushToken:String = defaults.string(forKey: "scg-push-token-dont-replace-this-default")! as String
        
        let params = ["app_id":appID,
                      "type":tokenType,
                      "token": pushToken] as Dictionary<String, String>
        
        let configuration = URLSessionConfiguration.default
        let session = URLSession(configuration: configuration)
        
        let urlString = "\(callbackURI)/push_tokens/unregister"
        let url = URL(string: urlString)
        
        var request = URLRequest(url: url!)
        request.httpMethod = "POST"
        request.timeoutInterval = 30
        
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.addValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        request.httpBody  = try! JSONSerialization.data(withJSONObject: params, options: [])
        
        let dataTask = session.dataTask(with: request, completionHandler: {
            (data: Data?, response: URLResponse?, error: Error?) -> Void in
            
            guard let httpResponse = response as? HTTPURLResponse, let _ = data
                else {
                    //print("error: not a valid http response")
                    if (failureBlock != nil) {
                        failureBlock! (error)
                    }
                    return
            }
            
            switch (httpResponse.statusCode)
            {
            case 200:
                defaults.removeObject(forKey: "scg-push-token-dont-replace-this-default")
                if (completionBlock != nil) {
                    completionBlock! ()
                }
            case 204:
                defaults.removeObject(forKey: "scg-push-token-dont-replace-this-default")
                if (completionBlock != nil) {
                    completionBlock! ()
                }
                
            default:
                if (failureBlock != nil) {
                    let errorSend = NSError(domain: (httpResponse.url?.absoluteString)!, code: httpResponse.statusCode, userInfo: nil)
                    failureBlock! (errorSend)
                }
            }
        })        

        dataTask.resume()
    }
    //[NSObject : AnyObject]
    open func deliveryConfirmation(userInfo:NSDictionary ,completionBlock: (() -> Void)? = nil, failureBlock : ((Error?) -> ())? = nil)
    {
        if let messageID = userInfo["scg-message-id"]
        {
            deliveryConfirmation(messageID as! String, completionBlock: completionBlock, failureBlock: failureBlock)
        }
    }
    
    open func deliveryConfirmation(_ messageID:String ,completionBlock: (() -> Void)? = nil, failureBlock : ((Error?) -> ())? = nil)
    {
        let configuration = URLSessionConfiguration.default
        let session = URLSession(configuration: configuration)
        
        let urlString = "\(callbackURI)/messages/\(messageID)/delivery_confirmation"
        let url = URL(string: urlString)
        
        var request = URLRequest(url: url!)
        request.httpMethod = "POST"
        request.timeoutInterval = 30
        
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.addValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        
        let dataTask = session.dataTask(with: request, completionHandler: {
            (data: Data?, response: URLResponse?, error: Error?) -> Void in
            
            guard let httpResponse = response as? HTTPURLResponse, let _ = data
                else {
                    if (failureBlock != nil) {
                        failureBlock! (error)
                    }
                    return
            }
            
            switch (httpResponse.statusCode)
            {
            case 200:
                if (completionBlock != nil) {
                    completionBlock! ()
                }
            case 204:
                if (completionBlock != nil) {
                    completionBlock! ()
                }
                
            default:
                if (failureBlock != nil) {
                    let errorSend = NSError(domain: (httpResponse.url?.absoluteString)!, code: httpResponse.statusCode, userInfo: nil)
                    failureBlock! (errorSend)
                }
            }
        })        

        dataTask.resume()
    }
    
    open func interactionConfirmation(userInfo:NSDictionary ,completionBlock: (() -> Void)? = nil, failureBlock : ((Error?) -> ())? = nil) {
        if let messageID = userInfo["scg-message-id"]
        {
            interactionConfirmation(messageID as! String, completionBlock: completionBlock, failureBlock: failureBlock)
        }
    }
    
    open func interactionConfirmation(_ messageID:String ,completionBlock: (() -> Void)? = nil, failureBlock : ((Error?) -> ())? = nil) {
        
        let configuration = URLSessionConfiguration.default
        let session = URLSession(configuration: configuration)
        
        let urlString = "\(callbackURI)/messages/\(messageID)/click_thru_confirmation"
        let url = URL(string: urlString)
        
        var request = URLRequest(url: url!)
        request.httpMethod = "POST"
        request.timeoutInterval = 30
        
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.addValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        
        let dataTask = session.dataTask(with: request, completionHandler: {
            (data: Data?, response: URLResponse?, error: Error?) -> Void in
            
            guard let httpResponse = response as? HTTPURLResponse, let _ = data
                else {
                    if (failureBlock != nil) {
                        failureBlock! (error)
                    }
                    return
            }
            
            switch (httpResponse.statusCode)
            {
            case 200:
                if (completionBlock != nil) {
                    completionBlock! ()
                }
            case 204:
                if (completionBlock != nil) {
                    completionBlock! ()
                }
                
            default:
                if (failureBlock != nil) {
                    let errorSend = NSError(domain: (httpResponse.url?.absoluteString)!, code: httpResponse.statusCode, userInfo: nil)
                    failureBlock! (errorSend)
                }
            }
        })
        
        dataTask.resume()
    }
    
    open func loadAttachment(userInfo:NSDictionary ,completionBlock: ((_ contentURL:URL, _ contentType:String) -> Void)? = nil, failureBlock : ((Error?) -> ())? = nil) {
        if let attachmentID = userInfo["scg-attachment-id"] as? String,
           let messageID = userInfo["scg-message-id"] as? String
        {
            self.loadAttachment(messageID, attachmentID: attachmentID, completionBlock: completionBlock, failureBlock: failureBlock)
        }
    }
    
    open func loadAttachment(_ messageID:String, attachmentID:String ,completionBlock: ((_ contentURL:URL, _ contentType:String) -> Void)? = nil, failureBlock : ((Error?) -> ())? = nil) {
        
        let configuration = URLSessionConfiguration.default
        let session = URLSession(configuration: configuration)
        let urlString = "\(callbackURI)/attachment/\(messageID)/\(attachmentID)"
//        let urlString = "https://www.hallaminternet.com/assets/https.jpg"
//        let urlString = "https://framework.realtime.co/blog/img/ios10-video.mp4"
        
        let url:URL = URL(string: urlString)!
        
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 30
        
        request.addValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        
        let dataTask = session.downloadTask(with: request, completionHandler: {
            (location: URL?, response: URLResponse?, error: Error?) -> Void in
            
            guard let httpResponse = response as? HTTPURLResponse
                else {
                    if (failureBlock != nil) {
                        failureBlock! (error)
                    }
                    return
            }
            var contentUrl:URL?
            let contentType:String = self.readTypeOfHeader(header: httpResponse)
            
            if let location = location {
                // Move temporary file to remove .tmp extension
                let tmpDirectory = NSTemporaryDirectory()
                let tmpFile = "file://".appending(tmpDirectory).appending(url.lastPathComponent)
                let tmpUrl = URL(string: tmpFile)!
                if FileManager.default.fileExists(atPath: tmpUrl.path) {
                    try! FileManager.default.removeItem(at: tmpUrl)
                }
                try! FileManager.default.moveItem(at: location, to: tmpUrl)
                contentUrl = tmpUrl
            }
            
            switch (httpResponse.statusCode)
            {
            case 200:
                if (completionBlock != nil) {
                    completionBlock! (contentUrl!, contentType)
                }
            case 204:
                if (completionBlock != nil) {
                    completionBlock! (contentUrl!, contentType)
                }
                
            default:
                if (failureBlock != nil) {
                    let errorSend = NSError(domain: (httpResponse.url?.absoluteString)!, code: httpResponse.statusCode, userInfo: nil)
                    failureBlock! (errorSend)
                }
            }
        })
        
        dataTask.resume()
    }
    
    open func saveDeviceToken(deviceTokenData tokenData:Data) {
        let tokenChars = (tokenData as NSData).bytes.bindMemory(to: CChar.self, capacity: tokenData.count)
        var pushToken = ""
        for i in 0..<tokenData.count {
            pushToken += String(format: "%02.2hhx", arguments: [tokenChars[i]])
            
        }
        
        let defaults = UserDefaults.standard
        defaults.set(pushToken, forKey: "scg-push-token-dont-replace-this-default")
    }
    
    open func saveDeviceToken(deviceToken token:NSString) {
        let defaults = UserDefaults.standard
        defaults.set(token, forKey: "scg-push-token-dont-replace-this-default")
    }
    
    func readTypeOfHeader(header:HTTPURLResponse) -> String {
        if let type:String = header.allHeaderFields["Content-Type"] as! String? {
            switch type {
            case "video/mpeg":
                return kUTTypeMPEG4 as String
            case "image/gif":
                return kUTTypeGIF as String
            case "image/png":
                return kUTTypePNG as String
            case "image/jpg":
                return kUTTypeImage as String
            default:
                return ""
            }
        }
        return ""
    }
}
